File: 2024/03_mull_it_over.py
import re

code = open('data/03.dat').read()

matches = re.findall('mul\([0-9]+,[0-9]+\)', code)

def multiply(match):
    a, b = map(int, match[4:-1].split(','))
    return a * b

print(sum(multiply(m) for m in matches))


matches = re.findall("mul\([0-9]+,[0-9]+\)|do\(\)|don't\(\)", code)

factor = 1
s = 0
for m in matches:
    if m == "do()":
        factor = 1
    elif m == "don't()":
        factor = 0
    else:
        s += factor * multiply(m)


print(s)

File: 2024/03_mull_it_over.jl
puzzle = read("data/03.dat", String)


function calculate_term(term)
    numbers = [parse(Int, s.match) for s in eachmatch(r"\d+", term)]
    return numbers[1] * numbers[2]
end

function solve2(matches)
    res, factor = 0, 1
    for term in matches
        if term == "do()"
            factor = 1
            continue
        end
        if term == "don't()"
            factor = 0
            continue
        end

        res += factor * calculate_term(term)
    end
    return res
end

@time begin
    matches = [s.match for s in eachmatch(r"(mul\(\d+,\d+\))", puzzle)]
    println("Part 1: ", sum(calculate_term(term) for term in matches))

    matches = [s.match for s in eachmatch(r"(mul\(\d+,\d+\))|(do\(\))|(don't\(\))", puzzle)]
    println("Part 2: ", solve2(matches))
end